src/Symfony/Component/HttpFoundation/IpRetriever/IpRetriever.php

```diff
@@ -0,0 +1,166 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpFoundation\IpRetriever;
+
+
+use Symfony\Component\HttpFoundation\IpUtils;
+use Symfony\Component\HttpFoundation\Request;
+
+/**
+ * IpRetriever
+ *
+ * Can retrieve real user ip from various contexts (proxified query or not).
+ * If you use a reverse proxy, set his ip with setTrustedProxies
+ * otherwise, you'll get your reverse proxy ip anyway.
+ *
+ * @author Xavier Leune <xavier.leune@gmail.com>
+ */
+class IpRetriever implements IpRetrieverInterface
+{
+    /**
+     * Names for headers that can be trusted when
+     * using trusted proxies.
+     *
+     * The FORWARDED header is the standard as of rfc7239.
+     *
+     * The other headers are non-standard, but widely used
+     * by popular reverse proxies (like Apache mod_proxy or Amazon EC2).
+     */
+    protected $trustedHeaders = array(
+        self::HEADER_FORWARDED => 'FORWARDED',
+        self::HEADER_CLIENT_IP => 'X_FORWARDED_FOR',
+        self::HEADER_REAL_IP => 'X_REAL_IP',
+    );
+
+    protected $trustedProxies = array();
+
+    /**
+     * {@inheritdoc}
+     */
+    public function setTrustedProxies(array $proxies)
+    {
+        $this->trustedProxies = $proxies;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getTrustedProxies()
+    {
+        return $this->getTrustedProxies();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function setTrustedHeaderName($key, $value)
```

Comment:
@stof can you take a look to this : https://github.com/symfony/symfony/compare/2.7...xavierleune:feature/ip-retriever-with-legacy-retriever?expand=1 ?
This could be a cleaner implementation with a LegacyIpRetriever, which have to be stateful because of the backward compatibility, but would be easily removed in sf3.
It you think it's the good way, I'll make changes here.

thx